File: 460 LFU Cache/solution.py
from collections import defaultdict, OrderedDict

class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.keyval = {}
        self.dictlist = defaultdict(lambda : OrderedDict())
        self._size = 0
        self._largest_freq = 0
            
        
    def get(self, key: int) -> int:
        # print("getting {}".format(key))
        # print(self.dictlist)
        # print(self.keyval)
        if key in self.keyval:
            val, freq = self.keyval[key]
            del self.dictlist[freq][key]
            self.keyval[key] = (val, freq + 1)
            self.dictlist[freq + 1][key] = val
            self._largest_freq = max(self._largest_freq, freq + 1)
            return val
        else:
            return -1
        

    def put(self, key: int, value: int) -> None:
        # print("putting {}".format(key))
        # print(self.dictlist)
        # print(self.keyval)
        if self.capacity == 0:
            return
        if key in self.keyval:
            _, freq = self.keyval[key]
            self.keyval[key] = (value, freq + 1)
            del self.dictlist[freq][key]
            self.dictlist[freq + 1][key] = value
            self._largest_freq = max(self._largest_freq, freq + 1)
        else:
            if self._size == self.capacity:
                for i in range(1, self._largest_freq + 1):
                    if len(self.dictlist[i]):
                        removed_key, _ = self.dictlist[i].popitem(last=False)
                        break
                del self.keyval[removed_key]
                self._size -= 1
            self.dictlist[1][key] = value
            self.keyval[key] = (value, 1)
            self._size += 1
            self._largest_freq = max(self._largest_freq, 1)


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 460 LFU Cache/solution.cpp
// Design and implement a data structure for Least Frequently Used (LFU) cache. It should support the following operations: get and put.

// get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
// put(key, value) - Set or insert the value if the key is not already present. When the cache reaches its capacity, it should invalidate the least frequently used item before inserting a new item. For the purpose of this problem, when there is a tie (i.e., two or more keys that have the same frequency), the least recently used key would be evicted.

// Follow up:
// Could you do both operations in O(1) time complexity?

// Example:

// LFUCache cache = new LFUCache( 2 /* capacity */ );

// cache.put(1, 1);
// cache.put(2, 2);
// cache.get(1);       // returns 1
// cache.put(3, 3);    // evicts key 2
// cache.get(2);       // returns -1 (not found)
// cache.get(3);       // returns 3.
// cache.put(4, 4);    // evicts key 1.
// cache.get(1);       // returns -1 (not found)
// cache.get(3);       // returns 3
// cache.get(4);       // returns 4
// TODO: implement this
#include <list>
#include <unordered_map>
#include <utility>

using namespace std;

class LFUCache {
    typedef unordered_map<int, unordered_map<int, int>::iterator> nodeMap;
    typedef list<pair<int, nodeMap> > LFUTYPE;
    const int cap;
    unordered_map<int, int> keyVal;
    LFUTYPE lfu;
    unordered_map<int, LFUTYPE::iterator> nodeFinder;

public:
    LFUCache(int capacity): cap(capacity) {
    }
    
    int get(int key) {
        if (keyVal.count(key) == 0)
            return -1;
        auto prevNode = nodeFinder[key];
        auto nodeIt = prevNode--;
        if ((*nodeIt).first)

        return keyVal[key];
    }
    
    void put(int key, int value) {
        
    }
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */